rust/src/measurements/make_private_lazyframe/group_by/mod.rs

```diff
@@ -21,136 +33,184 @@ use polars_plan::prelude::GroupbyOptions;
             let l1 = l0.inf_mul(&li)?.min(d_in);
-            Ok((l0, l1, li))
+
+            let mut d_out = privacy_map.eval(&(l0, l1, li))?;
+
+            if let Some((_, noise, threshold_value)) = &threshold {
+                let distribution = noise.distribution.expect("distribution is known");
+                let scale = noise.scale.expect("scale is known");
+
+                if li >= *threshold_value {
+                    return fallible!(FailedMap, "threshold must be greater than {:?}", li);
+                }
+
+                let d_instability = threshold_value.inf_sub(&li)?;
+                let delta_single =
+                    integrate_discrete_noise_tail(distribution, scale, d_instability)?;
+                let delta_joint =
+                    (1.0).inf_sub(&(1.0).inf_sub(&delta_single)?.inf_powi(IBig::from(l0))?)?;
+                d_out = MO::add_delta(d_out, delta_joint)?;
+            } else if margin.public_info.is_none() {
+                return fallible!(FailedMap, "keys must be public if threshold is unknown");
+            }
+
+            Ok(d_out)
         }),
     )?;
 
-    let m_exprs = (t_group_by
-        >> make_basic_composition(
-            aggs.into_iter()
-                .map(|expr| {
-                    make_private_expr(
-                        expr_domain.clone(),
-                        PartitionDistance(middle_metric.clone()),
-                        output_measure.clone(),
-                        expr.clone(),
-                        global_scale,
-                    )
-                })
-                .collect::<Fallible<_>>()?,
-        )?)?;
-
-    let f_exprs = m_exprs.function.clone();
-    let privacy_map = m_exprs.privacy_map.clone();
+    t_prior >> m_group_by_agg
+}
 
-    t_prior
-        >> Measurement::new(
-            middle_domain,
-            Function::new_fallible(move |arg: &DslPlan| {
-                let mut output = plan.clone();
-                if let DslPlan::GroupBy {
-                    ref mut input,
-                    ref mut aggs,
-                    ..
-                } = output
-                {
-                    *input = Arc::new(arg.clone());
-                    *aggs = f_exprs.eval(&(arg.clone(), all()))?;
-                };
-                Ok(output)
-            }),
-            middle_metric,
-            output_measure,
-            privacy_map,
-        )?
+pub trait ApproximateMeasure: BasicCompositionMeasure {
+    fn add_delta(d_out: Self::Distance, delta_p: f64) -> Fallible<Self::Distance>;
 }
 
-#[cfg(test)]
-mod test;
+macro_rules! impl_measure_non_catastrophic {
+    ($ty:ty) => {
+        impl ApproximateMeasure for $ty {
+            fn add_delta(_d_out: Self::Distance, _delta_p: f64) -> Fallible<Self::Distance> {
+                return fallible!(
+                    MakeMeasurement,
+                    "key-sets cannot be privatized under {:?}. Approximate-DP is necessary.",
```

Comment:
```suggestion
                    "key-sets cannot be privatized under {:?}. FixedSmoothedMaxDivergence is necessary.",
```

Not 100% sure that the suggestion is correct, but the idea that the suggested alternative is also a `*Divergence`... unless you think that will be obvious?